docs/Custom_Operation_for_GPUs.md

```diff
@@ -633,7 +633,73 @@ With this modification, the `all-gather` operation is eliminated and the custom
 
 ### Shard the forward function with custom_partitioning
 
+When using custom_partitioning, it isn't needed to define the JAX config
+experimental_xmap_spmd_lowering and
+experimental_xmap_spmd_lowering_manual.
 
+We create an helper function to help with all the JAX registration needed.
+
+```python
+def register_primitive(cls):
+...
+```
+
+We define 2 JAX primitives, one inner primitive that map to the
+real kernel we want to warp in JAX. And an outer primitive that will
+be used with the custom_partitioning registration and for the
+gradient. (And if you implement the interface to support vmat, it will
+also be on the outer primitive).
+
+JAX custom_partitioning implementation are callbacks from XLA to Python during XLA sharding logic.
+XLA sharding goes in two phases: a sharding propagation phase and a partition phase.
+The propagation phase is when XLA plan the sharding to be created. It is the partition phase that create the sharded graph.
+For XLA to be able to shard our custom operations, it needs us to define 2 extra functions:
+infer_sharding_from_operands() and partition(). They are used in the first and second phase respectively.
+
+The infer_sharding_from_operands() function must do what its name say: infer the output sharding from the input sharding.
+
+The partition() function will do a few things:
+- tell which input sharding will be expected. XLA will reshad if needed.
+- tell the final version of the output sharding.
+- give a function that will create the new instruction from the sharded inputs.
+
+See the code comments for more explanation:
+
+```python
+    def infer_sharding_from_operands(eps, mesh, arg_infos, result_infos):
```

Comment:
Some type annotations would be quite useful here, especially for people not entirely familiar with the signatures and expected arguments. For example:

`def infer_sharing_from_operands(
eps : float, mesh : jax.sharding.Mesh, arg_infos : <whatever the type>, result_infos : <whatever the type>
):`